ci/deploy/npm.py

```diff
@@ -0,0 +1,164 @@
+        package = json.load(package_file)
+    name = package["name"]
+    dist_tag = "dev" if version.is_development else "latest"
+    if version_exists_in_npm(name, version):
+        logger.warning("%s %s already released, skipping.", name, version.node)
+        return
+    else:
+        logger.info("Releasing %s %s", name, version.node)
+        set_npm_credentials(package_path)
+        spawn.runv(
+            ["npm", "publish", "--access", "public", "--tag", dist_tag],
+            cwd=package_path,
+        )
+
+
+def build_all(version: Version, *, do_release: bool = True) -> None:
+    for path_str in PROJECT_PATHS:
+        project_path = Path(path_str)
+        logger.info("Building %s @ %s", project_path, version.node)
+        package_path = build_package(version, project_path)
+        logger.info("Built %s", project_path)
+        if do_release:
+            release_package(version, package_path)
+            logger.info("Released %s", package_path)
+        else:
+            logger.info("Skipping release for %s", package_path)
+
+
+def version_exists_in_npm(name: str, version: Version) -> bool:
+    quoted = urllib.parse.quote(name)
+    res = requests.get(f"https://registry.npmjs.org/{quoted}/{version.node}")
+    if res.status_code == 404:
+        # This is a new package
+        return False
+    res.raise_for_status()
+    return True
+
+
+def set_npm_credentials(package_path: Path) -> None:
+    with open(package_path / ".npmrc", "w") as npmrc:
+        npmrc.write("//registry.npmjs.org/:_authToken=${NPM_TOKEN}\n")
+
+
+def parse_args() -> argparse.Namespace:
+    parser = argparse.ArgumentParser(
+        "npm.py", description="build and publish NPM packages"
+    )
+    parser.add_argument(
+        "-v,--verbose",
+        action="store_true",
+        dest="verbose",
+        help="Enable verbose logging",
+    )
+    parser.add_argument(
+        "--release",
+        action=argparse.BooleanOptionalAction,
+        dest="do_release",
+        default=True,
+        help="Whether or not the built package should be released",
+    )
+    parser.add_argument(
+        "--build-id",
+        type=int,
+        help="An optional build identifier. Used in pre-release version numbers",
+    )
+    return parser.parse_args()
+
+
+if __name__ == "__main__":
+    args = parse_args()
+    if args.verbose:
+        logger.setLevel(logging.DEBUG)
+    build_id = args.build_id
+    if os.environ.get("BUILDKITE_BUILD_NUMBER") is not None:
+        if build_id is not None:
+            logger.warning(
+                "Build ID specified via both envvar and CLI arg. Using CLI value"
+            )
+        else:
+            build_id = int(os.environ["BUILDKITE_BUILD_NUMBER"])
+    if args.do_release and "NPM_TOKEN" not in os.environ:
+        raise ValueError("'NPM_TOKEN' must be set")
+    crate_version = get_crate_version()
+    version = generate_version(crate_version, build_id)
```

Comment:
A couple suggestions above about creating a `cargo.Workspace`. Probably best to do that here, and then plumb it through to both `get_crate_version` and `build_all`. And then I think you might as well inline `get_crate_version` into `generate_version` because it'll be only one line.

Reply:
Thanks for this feedback - it's nice that all these helpers exist.